### PR TITLE
feat: add new transaction type MintTokenTx

### PR DESCRIPTION
**Pull request by Pangssu**
## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to manually override the Kroma Burgundy fork timestamp via a new command-line flag, enhancing user control over node configuration.
	- Added support for a new transaction type, `MintTokenTxType`, allowing for minting token transactions within the network.

- **Enhancements**
	- Enhanced transaction processing to accommodate the new `MintTokenTxType`, including adjustments in transaction validation, gas buying, and pre-check routines.
	- Updated RPC processing functions to handle mint token transactions, ensuring compatibility with external interfaces and applications.

- **Configuration**
	- Implemented a new field, `OverrideKromaBurgundy`, across various configurations and genesis settings to facilitate the override of the Kroma Burgundy fork settings.

### CODE REVIEW

**Review by sm-stack — APPROVED** [review]
Except for a small nit on comments, LGTM.

File: core/txpool/validation.go [hunk 1]
```diff
@@ -66,7 +66,7 @@ func ValidateTransaction(tx *types.Transaction, head *types.Header, signer types
 	// No unauthenticated deposits allowed in the transaction pool.
 	// This is for spam protection, not consensus,
 	// as the external engine-API user authenticates deposits.
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply core/txpool/validation.go:68]: Instead of modifying the content, we added the sentence below.

[the remark this answers is not in the document]
Comment (Pangssu) [reply core/txpool/validation.go:68]: Instead of modifying the content, I added the sentence below.

**Review by jyc228 — APPROVED** [review]
lgtm

File: core/types/mint_token_tx.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+const MintTokenTxType = 0x70
+
+type MintTokenTx struct {
+	From  common.Address  // From is always the mint function caller.
+	To    *common.Address // To always points to the MinterContract address.
```
Comment (kangsorang) [inline core/types/mint_token_tx.go:31]: If there's no clear purpose, wouldn't it be nice to unify the data types of From and To?

Comment (Pangssu) [reply core/types/mint_token_tx.go:31]: `To` field of the transaction type is pointer.

File: core/state_transition.go [hunk 3]
```diff
@@ -244,7 +246,7 @@ func (st *StateTransition) buyGas() error {
 	mgval := new(big.Int).SetUint64(st.msg.GasLimit)
 	mgval = mgval.Mul(mgval, st.msg.GasPrice)
 	var l1Cost *big.Int
-	if st.evm.Context.L1CostFunc != nil && !st.msg.SkipAccountChecks {
+	if st.evm.Context.L1CostFunc != nil && !st.msg.SkipAccountChecks && !st.msg.IsMintTokenTx {
 		l1Cost = st.evm.Context.L1CostFunc(st.evm.Context.BlockNumber.Uint64(), st.msg.RollupDataGas, st.msg.IsDepositTx)
```
Comment (0xHansLee) [inline core/state_transition.go:250]: In `NewL1CostFunc()` (`core/types/rollup_l1_cost.go`), do we need to check if the tx is mintTokenTx as well?
e.g.)
```go
if config.Kroma == nil || isDepositTx || isMintTokenTx || rollupDataGas == 0 {
	return nil
}
```

File: core/types/transaction_marshalling.go [hunk 4]
```diff
@@ -479,6 +485,40 @@ func (tx *Transaction) UnmarshalJSON(input []byte) error {
...
+		if (dec.V != nil && dec.V.ToInt().Cmp(common.Big0) != 0) ||
+			(dec.R != nil && dec.R.ToInt().Cmp(common.Big0) != 0) ||
+			(dec.S != nil && dec.S.ToInt().Cmp(common.Big0) != 0) {
+			return errors.New("mint token transaction signature must be 0 or unset")
+		}
+		var itx MintTokenTx
+		inner = &itx
+		if dec.To != nil {
+			itx.To = dec.To
+		}
```
Comment (seolaoh) [inline core/types/transaction_marshalling.go:505]:
```suggestion
		if dec.To == nil {
			return errors.New("missing required field 'to' in transaction")
		}
		itx.To = dec.To
```

File: core/types/transaction_marshalling.go [hunk 5]
```diff
@@ -479,6 +485,40 @@ func (tx *Transaction) UnmarshalJSON(input []byte) error {
 		if dec.Nonce != nil {
 			inner = &depositTxWithNonce{DepositTx: itx, EffectiveNonce: uint64(*dec.Nonce)}
 		}
+	case MintTokenTxType:
+		if dec.AccessList != nil || dec.MaxFeePerGas != nil ||
+			dec.MaxPriorityFeePerGas != nil {
+			return errors.New("unexpected field(s) in mint token transaction")
+		}
+		if dec.GasPrice != nil && dec.GasPrice.ToInt().Cmp(common.Big0) != 0 {
```
Comment (seolaoh) [inline core/types/transaction_marshalling.go:493]: Don't we need to check the existence of other unnecessary fields like `Value`, `Mint`, `SourceHash`?

File: core/state_transition.go (hunk 3 above)
Comment (seolaoh) [inline core/state_transition.go:250]: Isn't it better? Then `NewL1CostFunc` should be changed also.
```suggestion
	if st.evm.Context.L1CostFunc != nil && !st.msg.SkipAccountChecks {
		l1Cost = st.evm.Context.L1CostFunc(st.evm.Context.BlockNumber.Uint64(), st.msg.RollupDataGas, st.msg.IsDepositTx || st.msg.IsMintTokenTx)
```

File: core/state_transition.go [hunk 6]
```diff
@@ -473,7 +475,7 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 		// After EIP-3529: refunds are capped to gasUsed / 5
 		st.refundGas(params.RefundQuotientEIP3529)
 	}
-	if st.msg.IsDepositTx && rules.IsOptimismRegolith {
+	if st.msg.IsDepositTx && rules.IsOptimismRegolith || st.msg.IsMintTokenTx && rules.IsKromaBurgundy {
```
Comment (0xHansLee) [inline core/state_transition.go:478]: For better readability,
```suggestion
	if (st.msg.IsDepositTx && rules.IsOptimismRegolith) || (st.msg.IsMintTokenTx && rules.IsKromaBurgundy) {
```

**Review by 0xHansLee — COMMENTED** [review]
I left some comments that I think some parts missed.

1. In `core/vm/logger.go`, do we need to add condition in `MaybeAddFeeRecipientsToStatesAffected()`?

```go
func (l *StructLogger) MaybeAddFeeRecipientsToStatesAffected(tx *types.Transaction) {
	if !tx.IsDepositTx() && !tx.IsMintTokenTx() {
		l.statesAffected[params.KromaProtocolVault] = struct{}{}
		l.statesAffected[params.KromaProposerRewardVault] = struct{}{}
		l.statesAffected[params.KromaValidatorRewardVault] = struct{}{}
	}
}
```

**Comment by Pangssu** [comment]
Closing this PR because our team have made an internal decision to not create `MintTokenTx`, and the network upgrade is no longer needed.

**Pull request closed without merging**